commit 8fcf83857d43e526ce31af6ebedf1719f630ef77 (refs/changes/14/2414/2)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2017-08-18T01:40:43+00:00 (2 years, 2 months ago)
    
    joyent/node-cueball#123 ConnectionSet leak during failure
    Reviewed by: Cody Peter Mello <cody.mello@joyent.com>

diff --git a/lib/set.js b/lib/set.js
index 5e80905..62c8332 100644
--- a/lib/set.js
+++ b/lib/set.js
@@ -618,7 +618,7 @@ CueBallConnectionSet.prototype._incrCounter = function (counter) {
  *        |
  *        |
  *        v
- *    +--------+
+ *    +--------+ fsm failed
  *    |        | .drain()
  *    |  init  +-----------------------------+
  *    |        |                             |
@@ -704,6 +704,8 @@ LogicalConnection.prototype.state_init = function (S) {
 			if (self.lc_fsm.isInState('idle')) {
 				self.lc_hdl.try(self.lc_fsm);
 			}
+		} else if (st === 'failed' || st === 'cancelled') {
+			S.gotoState('stopped');
 		}
 	});
 
@@ -712,6 +714,8 @@ LogicalConnection.prototype.state_init = function (S) {
 			if (self.lc_hdl.isInState('waiting')) {
 				self.lc_hdl.try(self.lc_fsm);
 			}
+		} else if (st === 'failed') {
+			S.gotoState('stopped');
 		}
 	});
 
